### PR TITLE
Add a versatile nodeset generator

### LINKED ISSUE #5895: NodesetWithinDistance
Need a new MeshModifer that adds a nodeset to all nodes within a certain distance of a given point.  Syntax should look like:

```
[./nodeset]
  type = NodesetWithinDistance
  point = '0.2 0.3 0.1'
  distance = 2.7
  nodeset_id = 4
  nodeset_name = 'frog'
[../]
```

This would cause all nodes within `2.7` of the point to get added to the nodeset 4 named 'frog'.

### PR DESCRIPTION
**Pull request by GiudGiud**
closes https://github.com/idaholab/moose/issues/5895
refs https://github.com/idaholab/moose/issues/28881

To get around difficulties reported in #28869 using the Gmsh reader.
This is not the best motivation, but I can see this being useful regardless

### CODE REVIEW

**Comment by GiudGiud** [comment]
@pbehne would you have time to review? This is very close to what you did for sidesets

**Comment by pbehne** [comment]
> @pbehne would you have time to review? This is very close to what you did for sidesets

I'll take a look in the coming week!

**Comment by pbehne** [comment]
Starting on this now.

**Review by pbehne — CHANGES_REQUESTED** [review]

File: framework/include/meshgenerators/NodeSetsGeneratorBase.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+   * parameters
+   */
+  void setup(MeshBase & mesh);
+
+  /**
+   * Determines whether the node is on the exterior of the mesh
+   */
+  bool nodeOnMeshExteriorBoundary(const Node * node,
+                                  const std::vector<dof_id_type> & node_elems,
+                                  const MeshBase & mesh) const;
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/include/meshgenerators/NodeSetsGeneratorBase.h:37]: done

File: framework/include/meshgenerators/NodeSetsGeneratorBase.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  /**
+   * Determines whether the given node of an element belongs to any nodesets in the
+   * excluded_nodesets parameter.
+   */
+  bool nodeInExcludedNodesets(const std::vector<BoundaryID> & node_nodesets) const;
+
+  /**
+   * Determines whether the given element's node satisfies the user-specified constraints
+   */
+  bool nodeSatisfiesRequirements(const Node * node,
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/include/meshgenerators/NodeSetsGeneratorBase.h:61]: for sure

File: framework/src/meshgenerators/ParsedGenerateNodeset.C [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+std::unique_ptr<MeshBase>
+ParsedGenerateNodeset::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  setup(*mesh);
+
+  if (!mesh->is_replicated())
+    mooseError("Not implemented for distributed meshes");
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/meshgenerators/ParsedGenerateNodeset.C:89]: not tested in any MG afaik. Here we are hopeful someone will come in and implement it

File: framework/src/meshgenerators/ParsedGenerateSideset.C [hunk 4]
```diff
@@ -70,12 +68,7 @@ ParsedGenerateSideset::ParsedGenerateSideset(const InputParameters & parameters)
 
   // parse function
   if (_func_F->Parse(_function, "x,y,z") >= 0)
-    mooseError("Invalid function\n",
-               _function,
-               "\nin ParsedAddSideset ",
-               name(),
-               ".\n",
-               _func_F->ErrorMsg());
+    mooseError("Invalid function\n", _function, "\n", _func_F->ErrorMsg());
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/meshgenerators/ParsedGenerateSideset.C:71]: made consistent

File: framework/src/meshgenerators/ParsedGenerateNodeset.C [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    if (!nodeSatisfiesRequirements(curr_node, node_nodesets, node_elems, *mesh))
+      continue;
+
+    // Check expression
+    _func_params[0] = (*curr_node)(0);
+    _func_params[1] = (*curr_node)(1);
+    _func_params[2] = (*curr_node)(2);
+    if (evaluate(_func_F))
+    {
+      if (_replace)
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/meshgenerators/ParsedGenerateNodeset.C:124]: done

File: framework/src/meshgenerators/ParsedGenerateNodeset.C [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                      getParam<std::vector<std::string>>("constant_expressions"));
+
+  // parse function
+  if (_func_F->Parse(_function, "x,y,z") >= 0)
+    mooseError("Invalid function\n",
+               _function,
+               "\nin ParsedGenerateNodeset ",
+               name(),
+               ".\n",
+               _func_F->ErrorMsg());
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/meshgenerators/ParsedGenerateNodeset.C]: added

File: framework/src/meshgenerators/ParsedGenerateSideset.C (hunk 4 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/meshgenerators/ParsedGenerateSideset.C]: we could add a test. maybe in a dedicated PR on catching all MG exceptions missing coverage

File: framework/include/meshgenerators/NodeSetsGeneratorBase.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+   * subdomain_id_list.
+   */
+  bool nodeElementsInIncludedSubdomains(const std::vector<dof_id_type> node_elems,
+                                        const MeshBase & mesh) const;
+
+  /**
+   * Determines whether the given node of an element belongs to any nodesets in the
+   * included_nodesets parameter.
+   */
+  bool nodeInIncludedNodesets(const std::vector<BoundaryID> & node_nodesets) const;
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/include/meshgenerators/NodeSetsGeneratorBase.h:56]: done

**Review by pbehne — APPROVED** [review]

**Comment by GiudGiud** [comment]
thanks!
@roystgnr @lindsayad for the ccb review?

**Comment by GiudGiud** [comment]
or @cticenhour if you want newsletter content 📰

File: framework/include/meshgenerators/NodeSetsGeneratorBase.h [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  /// A list of included subdomain ids that the node has to be part of, extracted from the included_subdomains parameter
+  std::vector<subdomain_id_type> _included_subdomain_ids;
+
+  /// Whether to only include external node when considering nodes to add to the nodeset
+  const bool _include_only_external_nodes;
+
+  /// A map from nodes (ids) to  local elements (ids) which comprise the node
+  // Build the node to element map, which is usually provided by a MooseMesh but here we have a
+  // MeshBase.
+  std::map<dof_id_type, std::vector<dof_id_type>> _node_to_elem_map;
```
Comment (lindsayad) [inline framework/include/meshgenerators/NodeSetsGeneratorBase.h:110]:
```suggestion
  std::unordered_map<dof_id_type, std::vector<dof_id_type>> _node_to_elem_map;
```

**Review by lindsayad — APPROVED** [review]

**Comment by GiudGiud** [comment]
Test failures are unrelated.

Thanks for the reviews!

**Pull request merged**